jsondataferret/models.py

```diff
@@ -202,46 +208,125 @@ def get_data_fields(self):
+        """Returns dict of differences between this version of the record and the previous one"""
+        pcrh = self.get_previous_cached_record_history()
+        if pcrh:
+            return jsondiff.diff(pcrh.data, self.data)
+        else:
+            return self.data
+
+    def get_data_diff_previous_cached_record_history_html(self):
+        """HTML version of get_data_diff_previous_cached_record_history, for use in templates"""
+        return pygments.highlight(
+            json.dumps(self.get_data_diff_previous_cached_record_history(), indent=4),
+            pygments.lexers.data.JsonLexer(),
+            pygments.formatters.HtmlFormatter(),
+        )
+
     def get_data_fields_include_differences_from_latest_data(self):
+        """Returns list of fields with boolean key different_from_latest_value to indicate if they have changed from the current latest data.
+        This is intended to be used on edits not moderated yet, to show what the affect of approving them would be."""
         if self.approval_event or self.refusal_event:
             raise Exception(
                 "get_data_fields_include_differences_from_latest_data should only be used on edits not moderated yet"
             )
         # TODO work with data_key field.
-        if not self.record.cached_exists:
-            # No existing data, so nothing special we can do here
-            edit_fields = get_field_list_from_json(
-                self.record.type.public_id,
-                self.data,
+        return get_field_list_from_json_with_differences(
+            self.record.type.public_id,
+            self.record.cached_data if self.record.cached_exists else {},
+            self.data,
+            "different_from_latest_value",
+        )
+
+    def get_data_fields_include_differences_from_previous_data(self):
+        """Returns list of fields with boolean key different_from_previous_value to indicate if they have changed since this edit was created."""
+        # TODO work with data_key field.
+        previous_cached_record_history = self.get_previous_cached_record_history()
+        return get_field_list_from_json_with_differences(
+            self.record.type.public_id,
+            previous_cached_record_history.data
+            if previous_cached_record_history
+            else {},
+            self.data,
+            "different_from_previous_value",
+        )
+
+
+class CachedRecordHistory(models.Model):
+    """Stores the historical state of data on a record AFTER a particular event has been applied to it.
+    These are generated only after events with approved edits (created or refused edits do not change a record)
+    """
+
+    id = models.BigAutoField(primary_key=True)
+    record = models.ForeignKey(Record, on_delete=models.CASCADE)
+    event = models.ForeignKey(Event, on_delete=models.CASCADE)
+    data = models.JSONField(default=dict)
+    _previous_cached_record_history = -1
```

Comment:
To avoid the use of -1 here, you could not define it, and use hasattr to determine if it has been set.

Reply:
Done comment